File: libs/moqtail-rs/src/model/property/constant.rs
// Copyright 2025 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::convert::TryFrom;

use crate::model::error::ParseError;

#[repr(u64)]
#[derive(Clone, Debug, Copy, PartialEq)]
pub enum LOCPropertyId {
  Timestamp = 0x06,
  Timescale = 0x08,
  VideoFrameMarking = 0x0A,
  AudioLevel = 0x0C,
  VideoConfig = 0x0D,
}

impl TryFrom<u64> for LOCPropertyId {
  type Error = ParseError;

  fn try_from(value: u64) -> Result<Self, Self::Error> {
    match value {
      0x06 => Ok(LOCPropertyId::Timestamp),
      0x08 => Ok(LOCPropertyId::Timescale),
      0x0A => Ok(LOCPropertyId::VideoFrameMarking),
      0x0C => Ok(LOCPropertyId::AudioLevel),
      0x0D => Ok(LOCPropertyId::VideoConfig),
      _ => Err(ParseError::InvalidType {
        context: "LOCPropertyId::try_from(u64)",
        details: format!("Invalid type, got {value}"),
      }),
    }
  }
}

impl From<LOCPropertyId> for u64 {
  fn from(value: LOCPropertyId) -> Self {
    value as u64
  }
}

/// MOQT property type values.
/// Even type values use VarInt KVP encoding; odd type values use Bytes KVP encoding.
#[repr(u64)]
#[derive(Clone, Debug, Copy, PartialEq, Eq)]
pub enum TrackPropertyType {
  ObjectDeliveryTimeout = 0x02, // Track scope, VarInt (ms, 0 = no timeout)
  MaxCacheDuration = 0x04,      // Track scope, VarInt (ms)
  SubgroupDeliveryTimeout = 0x06, // Track scope, VarInt (ms, 0 = no timeout)
  ImmutableProperties = 0x0B,   // Track+Object scope, Bytes (nested KVPs)
  DefaultPublisherPriority = 0x0E, // Track scope, VarInt (0-255)
  DefaultPublisherGroupOrder = 0x22, // Track scope, VarInt (1=Ascending, 2=Descending)
  DynamicGroups = 0x30,         // Track scope, VarInt (0 or 1)
  PriorGroupIdGap = 0x3C,       // Object scope, VarInt
  PriorObjectIdGap = 0x3E,      // Object scope, VarInt
}

impl TryFrom<u64> for TrackPropertyType {
  type Error = ParseError;

  fn try_from(value: u64) -> Result<Self, Self::Error> {
    match value {
      0x02 => Ok(TrackPropertyType::ObjectDeliveryTimeout),
      0x04 => Ok(TrackPropertyType::MaxCacheDuration),
      0x06 => Ok(TrackPropertyType::SubgroupDeliveryTimeout),
      0x0B => Ok(TrackPropertyType::ImmutableProperties),
      0x0E => Ok(TrackPropertyType::DefaultPublisherPriority),
      0x22 => Ok(TrackPropertyType::DefaultPublisherGroupOrder),
      0x30 => Ok(TrackPropertyType::DynamicGroups),
      0x3C => Ok(TrackPropertyType::PriorGroupIdGap),
      0x3E => Ok(TrackPropertyType::PriorObjectIdGap),
      _ => Err(ParseError::InvalidType {
        context: "TrackPropertyType::try_from(u64)",
        details: format!("Unknown property type, got {value}"),
      }),
    }
  }
}

impl From<TrackPropertyType> for u64 {
  fn from(value: TrackPropertyType) -> Self {
    value as u64
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn test_track_property_type_roundtrip() {
    let known = [
      (0x02u64, TrackPropertyType::ObjectDeliveryTimeout),
      (0x04, TrackPropertyType::MaxCacheDuration),
      (0x06, TrackPropertyType::SubgroupDeliveryTimeout),
      (0x0B, TrackPropertyType::ImmutableProperties),
      (0x0E, TrackPropertyType::DefaultPublisherPriority),
      (0x22, TrackPropertyType::DefaultPublisherGroupOrder),
      (0x30, TrackPropertyType::DynamicGroups),
      (0x3C, TrackPropertyType::PriorGroupIdGap),
      (0x3E, TrackPropertyType::PriorObjectIdGap),
    ];
    for (raw, expected) in known {
      let parsed = TrackPropertyType::try_from(raw).unwrap();
      assert_eq!(parsed, expected);
      assert_eq!(u64::from(parsed), raw);
    }
  }

  #[test]
  fn test_track_property_type_unknown() {
    assert!(TrackPropertyType::try_from(0xFF).is_err());
    assert!(TrackPropertyType::try_from(0x00).is_err());
  }
}

File: libs/moqtail-ts/src/model/property/constant.ts
/**
 * Copyright 2025 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { InvalidTypeError } from '../error'

export enum LOCPropertyId {
  Timestamp = 0x06,
  Timescale = 0x08,
  VideoFrameMarking = 0x0a,
  AudioLevel = 0x0c,
  VideoConfig = 0x0d,
}

export enum TrackPropertyType {
  ObjectDeliveryTimeout = 0x02,
  MaxCacheDuration = 0x04,
  SubgroupDeliveryTimeout = 0x06,
  ImmutableProperties = 0x0b,
  DefaultPublisherPriority = 0x0e,
  DefaultPublisherGroupOrder = 0x22,
  DynamicGroups = 0x30,
  PriorGroupIdGap = 0x3c,
  PriorObjectIdGap = 0x3e,
}

export function locPropertyIdFromNumber(value: number): LOCPropertyId {
  switch (value) {
    case 0x06:
      return LOCPropertyId.Timestamp
    case 0x08:
      return LOCPropertyId.Timescale
    case 0x0a:
      return LOCPropertyId.VideoFrameMarking
    case 0x0c:
      return LOCPropertyId.AudioLevel
    case 0x0d:
      return LOCPropertyId.VideoConfig
    default:
      throw new InvalidTypeError('locPropertyIdFromNumber', `Invalid LOC property id: ${value}`)
  }
}

/**
 * @public
 * A registration-policy range in the Property Type space. `to` is absent for the
 * open-ended top range.
 */
export interface PropertyRange {
  readonly from: bigint
  readonly to?: bigint
}

/**
 * @public
 * Registration-policy ranges for the Property Type space (§15.8).
 *
 * §2.5 gives the 1-byte application-specific range as `0x38-0x3F`; that is stale text
 * from before draft-18's varint change, which made the 1-byte space `0x00-0x7F`.
 * §15.8 is the correct table and is what these constants follow.
 */
export const PropertyRanges = {
  /** Standards Action or IESG Approval (1-byte encoding). */
  StandardsAction: { from: 0x00n, to: 0x77n },
  /** Application-specific use, no registration permitted (1-byte encoding). */
  AppSpecific1Byte: { from: 0x78n, to: 0x7fn },
  /** Specification Required (2-byte encoding). */
  SpecRequired: { from: 0x80n, to: 0x37ffn },
  /** Application-specific use, no registration permitted (2-byte encoding). */
  AppSpecific2Byte: { from: 0x3800n, to: 0x3fffn },
  /** Mandatory Track Properties; Track scope only. */
  MandatoryTrack: { from: 0x4000n, to: 0x7fffn },
  /** First Come First Served begins here (open-ended). */
  Fcfs: { from: 0x8000n },
} as const satisfies Record<string, PropertyRange>

function inRange(range: PropertyRange, typeValue: bigint): boolean {
  return typeValue >= range.from && (range.to === undefined || typeValue <= range.to)
}

/**
 * @public
 * True if a Property Type is reserved for application-specific use (either encoding-width
 * range), for which no IANA registration is permitted.
 */
export function isApplicationSpecificProperty(typeValue: bigint | number): boolean {
  const v = BigInt(typeValue)
  return inRange(PropertyRanges.AppSpecific1Byte, v) || inRange(PropertyRanges.AppSpecific2Byte, v)
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest

  // Asserted against dev/conformance/draft18/, which is shared with moqtail-rs.
  describe('draft-18 conformance', () => {
    const fixture = async () => await import('../../../test/conformance')

    test('TrackPropertyType matches property_types.json', async () => {
      const { propertyTypes, assertRegistry, pascalIdent } = await fixture()
      assertRegistry(propertyTypes(), pascalIdent(), (codepoint) => TrackPropertyType[Number(codepoint)])
    })

    // The LOC properties are registered in the same number space as the draft's own
    // table (§15.8 Table 15), so they are held to it too.
    test('LOCPropertyId matches the provisional LOC registry', async () => {
      const { propertyTypes, assertRegistry, pascalIdent } = await fixture()
      assertRegistry(propertyTypes().provisional, pascalIdent(), (codepoint) => {
        try {
          return LOCPropertyId[locPropertyIdFromNumber(Number(codepoint))]
        } catch {
          return undefined
        }
      })
    })

    // The ranges, in order, including the open-ended First Come First Served one.
    test('PropertyRanges matches the ranges in property_types.json', async () => {
      const { propertyTypes, parseHex } = await fixture()
      const actual = Object.values(PropertyRanges).map((r) => [r.from, 'to' in r ? r.to : undefined])
      const expected = propertyTypes().ranges.entries.map((e) => [
        parseHex(e.from),
        e.to === null ? undefined : parseHex(e.to),
      ])
      expect(actual).toEqual(expected)
    })
  })

  describe('application-specific property ranges', () => {
    test('covers both encoding widths and nothing either side of them', () => {
      for (const v of [0x78n, 0x7fn, 0x3800n, 0x3fffn]) {
        expect(isApplicationSpecificProperty(v), `${v.toString(16)}`).toBe(true)
      }
      // 0x38-0x3F is §2.5's stale range and must not be treated as application-specific.
      for (const v of [0x38n, 0x3fn, 0x77n, 0x80n, 0x37ffn, 0x4000n]) {
        expect(isApplicationSpecificProperty(v), `${v.toString(16)}`).toBe(false)
      }
    })
  })
}
